src/server/game/Entities/Player/Anticheat.cpp

```diff
@@ -0,0 +1,396 @@
+                    sWorld->SendGMText(LANG_GM_ANNOUNCE_MOVE_UNDER_CONTROL, GetSession()->GetAccountId(), GetName().c_str());
+                    AccountMgr::RecordAntiCheatLog(GetSession()->GetAccountId(), GetName().c_str(), GetDescriptionACForLogs(4), GetPositionACForLogs(), int32(realm.Id.Realm));
+                    return false;
+                }
+            }
+        }
+
+        float distance, speed, difftime, normaldistance, delay, diffPacketdelay;
+        uint32 ptime;
+        std::string mapname = GetMap()->GetMapName();
+
+        // calculate distance - don't use func, because x,z can be offset transport coords
+        distance = sqrt((npos.GetPositionY() - y) * (npos.GetPositionY() - y) + (npos.GetPositionX() - x) * (npos.GetPositionX() - x));
+        
+        if (!jump && !CanFly() && !isSwimming() && !transportflag)
+        {
+            float diffz = fabs(movementInfo.pos.GetPositionZ() - z);
+            float tanangle = distance / diffz;
+
+            if (movementInfo.pos.GetPositionZ() > z &&
+                diffz > 1.87f &&
+                tanangle < 0.57735026919f) // 30 degrees
+            {
+                TC_LOG_INFO("anticheat", "Player::CheckMovementInfo :  Climb-Hack detected for Account id : %u, Player %s, diffZ = %f, distance = %f, angle = %f, Map = %s, mapId = %u, X = %f, Y = %f, Z = %f",
+                    GetSession()->GetAccountId(), GetName().c_str(), diffz, distance, tanangle, mapname.c_str(), GetMapId(), x, y, z);
+                sWorld->SendGMText(LANG_GM_ANNOUNCE_WALLCLIMB, GetSession()->GetAccountId(), GetName().c_str(), diffz, distance, tanangle, mapname.c_str(), GetMapId(), x, y, z);
+                AccountMgr::RecordAntiCheatLog(GetSession()->GetAccountId(), GetName().c_str(), GetDescriptionACForLogs(5, diffz, distance), GetPositionACForLogs(), int32(realm.Id.Realm));
+                return false;
+            }
+        }
+
+        uint32 oldstime = GetLastMoveServerTimestamp();
+        uint32 stime = GameTime::GetGameTimeMS();
+        uint32 ping;
+        ptime = movementInfo.time;
+
+        if (!vehicle)
+            speed = GetSpeed(MOVE_RUN);
+        else
+            speed = GetVehicleKit()->GetBase()->GetSpeed(MOVE_RUN);
+        if (isSwimming())
+        {
+            if (!vehicle)
+                speed = GetSpeed(MOVE_SWIM);
+            else
+                speed = GetVehicleKit()->GetBase()->GetSpeed(MOVE_SWIM);
+        }
+        if (IsFlying() || GetPlayerMovingMe()->CanFly())
+        {
+            if (!vehicle)
+                speed = GetSpeed(MOVE_FLIGHT);
+            else
+                speed = GetVehicleKit()->GetBase()->GetSpeed(MOVE_FLIGHT);
+        }
+
+        delay = ptime - oldctime;
+        diffPacketdelay = 10000000 - delay;
+
+        if (oldctime > ptime)
+        {
+            TC_LOG_INFO("anticheat", "oldctime > ptime");
+            delay = 0;
+        }
+        diffPacketdelay = diffPacketdelay * 0.0000000001f;
```

Comment:
what is the deal with all these 10000000 and 0.0000000001f numbers ? what are they supposed to do ?